### PR TITLE
Allow elasticity in room joining 

### PR DESCRIPTION
**Pull request by ah1053**
1. Added MinPlayersToStartRoom as part of Room class, JoinOrCreateRoom Request and CreateRoom Request 
> Addresses https://github.com/mobiledgex/edge-multiplay-node-server/issues/13#issue-875349430.

2. New Event (Rooms Updated) notification for the following:
- Player joins a room.
- Player leaves a room.

3. Tests update:
- New testing matrix
- Added tests for MinPlayersToStartGame and CloseServer method.

4. Added new public method closeServer() for closing all the running servers.
5. Added an instanceOf override for  GamePlayEvent class. 

6. Updated Docker file image to use node:16-slim ([official node dockerhub](https://hub.docker.com/_/node))

### CODE REVIEW

File: models/Events.js [hunk 1]
```diff
@@ -212,6 +212,41 @@ class GamePlayEvent extends Event {
...
+    if (obj.roomId === "") {
+      console.error("missing room id");
+      return false;
+    }
+    if (obj.eventName === "") {
+      console.error("missing event name");
+      return false;
+    }
+    if (typeof (obj.stringData) === Array) {
+      console.error("stringData is not an array");
```
Comment (brucearmstrong) [inline models/Events.js:233]: Checks that it "is" an array, but logs that it "is not" an array.

Comment (lgarner) [reply models/Events.js:233]: Besides that, null is possible. typeData seems quite raw as a generic type. Why not Json object array?

File: server.js [hunk 2]
```diff
@@ -251,6 +254,11 @@ function startEdgeMultiplay () {
     try {
       var gameplayEventStr = new Buffer.from(gameplayEventBinary).toString();
       var jsonObj = JSON.parse(gameplayEventStr);
+      if (jsonObj instanceof Events.GamePlayEvent === false) {
+        console.log("Received unkonw event", jsonObj);
```
Comment (brucearmstrong) [inline server.js:258]: nit: "unkonw" -> "unknown"

File: server.js [hunk 3]
```diff
@@ -315,6 +323,25 @@ function startEdgeMultiplay () {
...
   });
 
+  udpServer.on('close', () => {
+    console.log("UDP Server closed");
+  });
+
+  statsServer.on('close', () => {
+    console.log("Stats Server closed");
+  });
+  wsStatsServer.on("close", () => {
```
Comment (brucearmstrong) [inline server.js:333]: nit: newline to match other code.

**Review by lgarner — APPROVED** [review]
👍

File: models/Room.js [hunk 4]
```diff
@@ -25,19 +25,23 @@ class Room {
     /**
      * @constructor
      * @param  {integer} maxPlayersPerRoom maximum players per room, must be greater than 1
+     * @param  {integer} minPlayersToStartGame minimum players to start the game, more players can join afterwards as long as the number of players in a room is less than maxPlayersPerRoom, when minPlayersToStartGame is less than 2 minPlayersToStartGame will be equal maxPlayersPerRoom 
      */
-    constructor (maxPlayersPerRoom) {
+    constructor (maxPlayersPerRoom, minPlayersToStartGame = 0) {
         this.roomId = uuidv4();
         this.roomMembers = [];
+        this.minPlayersToStartGame = minPlayersToStartGame < 2 ? maxPlayersPerRoom : minPlayersToStartGame;
```
Comment (lgarner) [inline models/Room.js:33]: Nit: Hard coded const.

**Review by brucearmstrong — APPROVED** [review]

**Review by brucearmstrong — APPROVED** [review]

File: models/Events.js (hunk 1 above)
Comment (ah1053) [reply models/Events.js:233]: Thanks for this, I fixed the conditions and updated the conditions to allow undefined (null) values for stringData, booleanData, intData .
Also added some unit tests to verify that.
36afed8

File: models/Room.js [hunk 5]
```diff
@@ -68,6 +73,10 @@ class Room {
         return this.roomMembers.length === this.maxPlayersPerRoom;
     }
 
+    isReadyToStartGame () {
+        return this.roomMembers.length === this.minPlayersToStartGame;
```
[the remark this answers is not in the document]
Comment (ah1053) [reply models/Room.js]: Removed GameStart from the server so I removed this property also

File: server.js (hunk 2 above)
Comment (ah1053) [reply server.js]: Fixed

**Pull request merged**